p2p/peerinfo/peerinfo.go

```diff
@@ -0,0 +1,389 @@
+package peerinfo
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/protocol"
+	"github.com/libp2p/go-libp2p/p2p/protocol/holepunch"
+	ma "github.com/multiformats/go-multiaddr"
+	"golang.org/x/exp/maps"
+	"golang.org/x/sync/errgroup"
+)
+
+type Kind string
+
+const (
+	KindUknown            Kind = ""
+	KindInbound           Kind = "inbound"
+	KindOutbound          Kind = "outbound"
+	KindHolePunchInbound  Kind = "hp_inbound"
+	KindHolePunchOutbound Kind = "hp_outbound"
+	KindHolePunchUnknown  Kind = "hp_unknown"
+	KindRelayInbound      Kind = "relay_in"
+	KindRelayOutbound     Kind = "relay_out"
+
+	bpsInterval1 = 10 * time.Second
+	bpsInterval2 = 5 * time.Minute
+
+	totalProto = "__total__"
+	otherProto = "__other__"
+)
+
+type PeerRequestStats struct {
+	sync.Mutex
+	successCount int
+	failureCount int
+	duration     time.Duration
+}
+
+func (ps *PeerRequestStats) SuccessCount() int {
+	ps.Lock()
+	defer ps.Unlock()
+	return ps.successCount
+}
+
+func (ps *PeerRequestStats) FailureCount() int {
+	ps.Lock()
+	defer ps.Unlock()
+	return ps.failureCount
+}
+
+func (ps *PeerRequestStats) Latency() time.Duration {
+	ps.Lock()
+	defer ps.Unlock()
+	count := ps.successCount + ps.failureCount
+	if count == 0 {
+		return 0
+	}
+	return ps.duration / time.Duration(count)
+}
+
+func (ps *PeerRequestStats) RequestDone(took time.Duration, success bool) {
+	ps.Lock()
+	defer ps.Unlock()
+	if success {
+		ps.successCount++
+	} else {
+		ps.failureCount++
+	}
+	ps.duration += took
+}
+
+type DataStats struct {
+	sync.Mutex
```

Comment:
This shouldn't be public in my opinion, or anyone with access to the `DataStats` object can lock and unlock the contained mutex.

Reply:
Done